codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -126,297 +118,189 @@ public void decode(ByteBuf in, Http2Headers headers) throws IOException {
 
-                    // Check for numerical overflow
-                    if (valueLength > Integer.MAX_VALUE - index) {
+                    if (valueLength < 0) { // Check for numerical overflow
                         throw DECODE_DECOMPRESSION_EXCEPTION;
                     }
-                    valueLength += index;
 
                     // Check new header size against max header size
-                    long newHeaderSize = (long) nameLength + (long) valueLength;
-                    if (newHeaderSize + headerSize > maxHeaderSize) {
-                        // truncation will be reported during endHeaderBlock
-                        headerSize = maxHeaderSize + 1;
-
-                        if (indexType == IndexType.NONE) {
-                            // Value is unused so skip bytes
-                            state = State.SKIP_LITERAL_HEADER_VALUE;
-                            break;
-                        }
-
-                        // Check new header size against max dynamic table size
-                        if (newHeaderSize + HeaderField.HEADER_ENTRY_OVERHEAD > dynamicTable.capacity()) {
-                            dynamicTable.clear();
-                            state = State.SKIP_LITERAL_HEADER_VALUE;
-                            break;
-                        }
+                    if ((long) valueLength + nameLength > maxHeadersLength - headersLength) {
+                        maxHeaderSizeExceeded();
                     }
-                    state = State.READ_LITERAL_HEADER_VALUE;
+                    state = READ_LITERAL_HEADER_VALUE;
                     break;
 
                 case READ_LITERAL_HEADER_VALUE:
                     // Wait until entire value is readable
                     if (in.readableBytes() < valueLength) {
-                        return;
+                        throw notEnoughDataException();
                     }
 
-                    CharSequence value = readStringLiteral(in, valueLength);
-                    insertHeader(headers, name, value, indexType);
-                    state = State.READ_HEADER_REPRESENTATION;
-                    break;
-
-                case SKIP_LITERAL_HEADER_VALUE:
-                    int skipBytes = min(in.readableBytes(), valueLength);
-                    in.skipBytes(skipBytes);
-                    valueLength -= skipBytes;
-                    if (valueLength == 0) {
-                        state = State.READ_HEADER_REPRESENTATION;
-                    }
+                    CharSequence value = readStringLiteral(in, valueLength, huffmanEncoded);
+                    headersLength = insertHeader(headers, name, value, indexType, headersLength);
+                    state = READ_HEADER_REPRESENTATION;
                     break;
 
                 default:
-                    throw new IllegalStateException("should not reach here");
+                    throw new IllegalStateException("should not reach here state: " + state);
```

Comment:
Micronit: Technically IllegalStateException is for people calling methods out of order (like calling close() and then read())

I don't have a better exception to use though :/